sessions-core/src/main/java/com/adyen/checkout/sessions/api/SessionService.kt

```diff
@@ -0,0 +1,132 @@
+
+internal class SessionService(
+    private val baseUrl: String,
+) {
+
+    suspend fun setupSession(
+        request: SessionSetupRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionSetupResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/setup?clientKey=$clientKey",
+            requestJson = SessionSetupRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionSetupResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    suspend fun submitPayment(
+        request: SessionPaymentsRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionPaymentsResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/payments?clientKey=$clientKey",
+            requestJson = SessionPaymentsRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionPaymentsResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    suspend fun submitDetails(
+        request: SessionDetailsRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionDetailsResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/paymentDetails?clientKey=$clientKey",
+            requestJson = SessionDetailsRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionDetailsResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    suspend fun checkBalance(
+        request: SessionBalanceRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionBalanceResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/paymentMethodBalance?clientKey=$clientKey",
+            requestJson = SessionBalanceRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionBalanceResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    suspend fun createOrder(
+        request: SessionOrderRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionOrderResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/orders?clientKey=$clientKey",
+            requestJson = SessionOrderRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionOrderResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    suspend fun cancelOrder(
+        request: SessionCancelOrderRequest,
+        sessionId: String,
+        clientKey: String,
+    ): SessionCancelOrderResponse = withContext(Dispatchers.IO) {
+        val resultJson = makeRequest(
+            path = "v1/sessions/$sessionId/orders/cancel?clientKey=$clientKey",
+            requestJson = SessionCancelOrderRequest.SERIALIZER.serialize(request),
+        )
+
+        SessionCancelOrderResponse.SERIALIZER.deserialize(resultJson)
+    }
+
+    private fun makeRequest(
```

Comment:
We could make this even more generic by creating a global extension function every service could use. Something like:
```
fun <T, R> HttpClient.post(
    baseUrl: String,
    path: String,
    request: T,
    requestSerializer: Serializer<T>,
    responseSerializer: Serializer<R>,
): R {
    Logger.v(TAG, "call - $path")
    val requestJson = requestSerializer.serialize(request)
    Logger.v(TAG, "request - ${requestJson.toStringPretty()}")
    val httpClient = HttpClientFactory.getHttpClient(baseUrl)
    val result = httpClient.post(path, requestJson.toString())
    val resultJson = JSONObject(String(result, Charsets.UTF_8))
    Logger.v(TAG, "response: ${resultJson.toStringPretty()}")
    return responseSerializer.deserialize(resultJson)
}
```
What do you think?

Reply:
I like the idea! Maybe make a new commit with it and we can discuss it there?

Reply:
https://github.com/Adyen/adyen-android/pull/655/commits/1255ac1c51c3cb8b4f70fc0a4c0793908071844e